core/directory.go

```diff
@@ -523,6 +523,79 @@ func (dir *Directory) WithPatch(ctx context.Context, patch string) (*Directory,
 	return dir, nil
 }
 
+func (dir *Directory) Search(ctx context.Context, opts SearchOpts, paths []string, globs []string) ([]*SearchResult, error) {
+	// Validate and normalize paths to prevent directory traversal attacks
+	for i, p := range paths {
+		// If absolute, make it relative to the directory
+		if filepath.IsAbs(p) {
+			paths[i] = strings.TrimPrefix(p, "/")
+		}
+
+		// Clean the path (e.g., remove ../, ./, etc.)
+		paths[i] = filepath.Clean(paths[i])
+
+		// Check if the normalized path would escape the directory
+		if !filepath.IsLocal(paths[i]) {
+			return nil, fmt.Errorf("path cannot escape directory: %s", p)
+		}
+	}
+
+	ref, err := getRefOrEvaluate(ctx, dir)
+	if err != nil {
+		return nil, err
+	}
+	if ref == nil {
+		// empty directory, i.e. llb.Scratch()
+		return []*SearchResult{}, nil
+	}
+
+	opt, ok := buildkit.CurrentOpOpts(ctx)
+	if !ok {
+		return nil, fmt.Errorf("no buildkit opts in context")
+	}
+	ctx = trace.ContextWithSpanContext(ctx, opt.CauseCtx)
+
+	bkSessionGroup, ok := buildkit.CurrentBuildkitSessionGroup(ctx)
+	if !ok {
+		return nil, fmt.Errorf("no buildkit session group in context")
+	}
+
+	results := []*SearchResult{}
+	err = MountRef(ctx, ref, bkSessionGroup, func(root string) error {
+		resolvedDir, err := containerdfs.RootPath(root, dir.Dir)
+		if err != nil {
+			return err
+		}
+		rgArgs := opts.RipgrepArgs()
+		for _, glob := range globs {
+			rgArgs = append(rgArgs, "--glob="+glob)
+		}
+		if len(paths) > 0 {
+			rgArgs = append(rgArgs, "--")
+			for _, p := range paths {
+				resolved, err := containerdfs.RootPath(root, p)
+				if err != nil {
+					return err
+				}
+				// make it relative, now that it's safe, just for less obtuse errors
+				resolved, err = filepath.Rel(root, resolved)
+				if err != nil {
+					return err
+				}
```

Comment:
```suggestion
				resolved, err := containerdfs.RootPath(resolvedDir, p)
				if err != nil {
					return err
				}
				// make it relative, now that it's safe, just for less obtuse errors
				resolved, err = filepath.Rel(resolvedDir, resolved)
				if err != nil {
					return err
				}
```
I think? Just so that we make path relative to `dir.Dir` and not the snapshot's root, since `dir.Dir` is what the caller sees as dir's "root"

Reply:
My brain hurts now, but managed to test it + apply the fix - great catch!